examples/common/convert_c4.py

```diff
@@ -4,105 +4,344 @@
+        )
+
+    # Make sure we have needed concat options
+    if (parsed.concat_text is not None and
+            isinstance(parsed.concat_text, int) and parsed.bos_text is None and
+            parsed.eos_text is None):
+        parser.error(
+            dedent(
+                'When setting --concat_text, you must specify at least one of --bos_text or --eos_text \
+            with which to separate concatenated sequences'))
+    if (parsed.concat_tokens is not None and
+            isinstance(parsed.concat_tokens, int) and parsed.tokenizer is None):
+        parser.error(
+            'When setting --concat_tokens, you must specify a --tokenizer')
+
+    # now that we have validated them, change BOS/EOS to strings
+    if parsed.bos_text is None:
+        parsed.bos_text = ''
+    if parsed.eos_text is None:
+        parsed.eos_text = ''
+    return parsed
+
+
+class NoConcatC4(IterableDataset):
+    """An IterableDataset that returns text samples for MDSWriter.
+
+    Returns dicts of {'text': bytes}
+    """
+
+    def __init__(self, split: str):
+        self.hf_dataset = hf_datasets.load_dataset(path='c4',
+                                                   name='en',
+                                                   split=split,
+                                                   streaming=True)
 
-    return args.parse_args()
+    def __iter__(self) -> Iterable[Dict[str, bytes]]:
+        for sample in self.hf_dataset:
+            # convert to bytes to store in MDS binary format
+            yield {'text': sample['text'].encode('utf-8')}
 
 
-def build_hf_c4_dataset(split: str) -> IterableDataset:
-    """Collect the samples for this dataset split.
+class ConcatTextC4(IterableDataset):
+    """An IterableDataset that returns text samples for MDSWriter.
+
+    Returns dicts of {'text': bytes}
+    """
+
+    def __init__(self, split: str, max_length: int, bos_text: str,
+                 eos_text: str, no_wrap: bool):
+        self.max_length = max_length
+        self.bos_text = bos_text
+        self.eos_text = eos_text
+        self.should_wrap = not no_wrap
+        self.hf_dataset = hf_datasets.load_dataset(path='c4',
+                                                   name='en',
+                                                   split=split,
+                                                   streaming=True)
+
+    def __iter__(self) -> Iterable[Dict[str, bytes]]:
+        buffer = ''
+        for sample in self.hf_dataset:
+            buffer = buffer + self.bos_text + sample['text'] + self.eos_text
+            while len(buffer) >= self.max_length:
+                concat_sample = buffer[:self.max_length]
+                buffer = buffer[self.max_length:] if self.should_wrap else ''
+                # convert to bytes to store in MDS binary format
+                yield {'text': concat_sample.encode('utf-8')}
+
+
+class ConcatTokensC4(IterableDataset):
```

Comment:
Should inherit from `ConcatTextC4`.

Reply:
I really don't think the inheritance does much here... like if I'm an external person and I just want the code for concatenating tokens, easy to just rip this class. Rather than rip 3 classes

Reply:
yeah, i guess you'd get a few items shared in the constructor, but that is it. I think I will agree with abhi and leave as is for now.